File: packages/ts-host/src/parse.ts
import type { RequestContext } from '@a2a-js/sdk/server'
import type { Ai37Metadata, A2uiAction } from './types'

export interface ParsedMessage {
  text?: string
  data: Record<string, unknown>
  metadata: Ai37Metadata
  /** A2UI-действие (клик/submit), если оркестратор форварднул его в `message.metadata.a2uiAction`. */
  action?: A2uiAction
  /**
   * W3C trace-context из `message.metadata` (`{ traceparent, tracestate? }`), который оркестратор
   * прокинул через `injectTraceContext`. Host продолжает по нему распределённый трейс (Langfuse v4).
   */
  traceCarrier?: Record<string, string>
}

/** Нормализует A2A-сообщение: текст + data-part + конверт metadata.ai37 + A2UI-действие. */
export function parseA2AMessage(rc: RequestContext): ParsedMessage {
  const parts = rc.userMessage.parts
  const textPart = parts.find((p) => p.kind === 'text')
  const dataPart = parts.find((p) => p.kind === 'data')
  const text = textPart?.kind === 'text' ? textPart.text : undefined
  const data = (dataPart?.kind === 'data' ? dataPart.data : {}) as Record<
    string,
    unknown
  >
  const action = readA2uiAction(rc)
  const traceCarrier = readTraceCarrier(rc)
  return {
    text,
    data,
    metadata: readAi37Metadata(rc, data),
    ...(action ? { action } : {}),
    ...(traceCarrier ? { traceCarrier } : {}),
  }
}

/**
 * W3C trace-context из `message.metadata` (`traceparent`/`tracestate`), который оркестратор положил
 * туда через `injectTraceContext` при исходящем A2A-вызове. undefined, если оркестратор не трассирует.
 */
function readTraceCarrier(rc: RequestContext): Record<string, string> | undefined {
  const md = rc.userMessage.metadata as Record<string, unknown> | undefined
  const traceparent = md?.traceparent
  if (typeof traceparent !== 'string' || traceparent.length === 0) return undefined
  const carrier: Record<string, string> = { traceparent }
  if (typeof md?.tracestate === 'string') carrier.tracestate = md.tracestate
  return carrier
}

/**
 * A2UI-действие из `message.metadata.a2uiAction.userAction` — зеркало AG-UI-пути
 * (`agui.ts readA2uiAction` читает `forwardedProps.a2uiAction.userAction`). Так оркестратор
 * форвардит клик/submit формы вниз конечному агенту по A2A. `name:string` обязателен,
 * `context` → `{}` по умолчанию. Нет действия (обычный текстовый ход) → undefined.
 */
function readA2uiAction(rc: RequestContext): A2uiAction | undefined {
  const ua = (
    rc.userMessage.metadata as { a2uiAction?: { userAction?: unknown } } | undefined
  )?.a2uiAction?.userAction as
    | { name?: unknown; context?: unknown; surfaceId?: unknown; sourceComponentId?: unknown }
    | undefined
  if (!ua || typeof ua.name !== 'string') return undefined
  const action: A2uiAction = {
    name: ua.name,
    context: (ua.context as Record<string, unknown> | undefined) ?? {},
  }
  if (typeof ua.surfaceId === 'string') action.surfaceId = ua.surfaceId
  if (typeof ua.sourceComponentId === 'string') action.sourceComponentId = ua.sourceComponentId
  return action
}

/** metadata.ai37 может прийти в message.metadata, data.ai37 или data.metadata.ai37. */
function readAi37Metadata(
  rc: RequestContext,
  data: Record<string, unknown>,
): Ai37Metadata {
  const fromMsg = (
    rc.userMessage.metadata as Record<string, unknown> | undefined
  )?.ai37 as Ai37Metadata | undefined
  const nested = (data.metadata as Record<string, unknown> | undefined)?.ai37
  const fromData = (data.ai37 ?? nested) as Ai37Metadata | undefined
  return { ...(fromData ?? {}), ...(fromMsg ?? {}) }
}

File: packages/python-host/src/ai37_agent_host/parse.py
"""Нормализация входящего A2A-сообщения — порт ``ts-host/src/parse.ts``.

``a2a-sdk`` 1.x — protobuf-мир: ``rc.message`` — это ``a2a_pb2.Message``. Читаем его через
``MessageToDict`` (как Minstroy ``_extract_payload``), затем достаём text/data-парты, конверт
``metadata.ai37``, A2UI-действие и W3C trace-carrier.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from google.protobuf.json_format import MessageToDict

from .output_modes import read_client_capabilities
from .types import A2uiAction, Ai37Metadata, ContextFile, IntentEnvelope


@dataclass
class ParsedMessage:
    data: dict[str, Any]
    metadata: Ai37Metadata
    text: str | None = None
    #: A2UI-действие (клик/submit) из message.metadata.a2uiAction (форвард оркестратора).
    action: A2uiAction | None = None
    #: W3C trace-context (traceparent/tracestate) из message.metadata (injectTraceContext).
    trace_carrier: dict[str, str] | None = None
    #: supportedCatalogIds клиента (a2uiClientCapabilities.v0.9) из метаданных сообщения.
    supported_catalog_ids: list[str] = field(default_factory=list)


def parse_a2a_message(rc: Any) -> ParsedMessage:
    """Нормализует A2A-сообщение: текст + data-part + конверт metadata.ai37 + A2UI-действие."""
    message = getattr(rc, "message", None)
    msg = MessageToDict(message, preserving_proto_field_name=False) if message is not None else {}
    parts = msg.get("parts") or []
    text: str | None = None
    data: dict[str, Any] = {}
    for part in parts:
        if not isinstance(part, dict):
            continue
        if text is None and isinstance(part.get("text"), str):
            text = part["text"]
        elif not data and isinstance(part.get("data"), dict):
            data = part["data"]
    md = msg.get("metadata") or {}
    return ParsedMessage(
        text=text,
        data=data,
        metadata=_read_ai37_metadata(md, data),
        action=_read_a2ui_action(md),
        trace_carrier=_read_trace_carrier(md),
        supported_catalog_ids=read_client_capabilities(md),
    )


def _read_trace_carrier(md: dict[str, Any]) -> dict[str, str] | None:
    traceparent = md.get("traceparent")
    if not isinstance(traceparent, str) or not traceparent:
        return None
    carrier = {"traceparent": traceparent}
    tracestate = md.get("tracestate")
    if isinstance(tracestate, str):
        carrier["tracestate"] = tracestate
    return carrier


def _read_a2ui_action(md: dict[str, Any]) -> A2uiAction | None:
    envelope = md.get("a2uiAction")
    user_action = envelope.get("userAction") if isinstance(envelope, dict) else None
    if not isinstance(user_action, dict) or not isinstance(user_action.get("name"), str):
        return None
    ctx = user_action.get("context")
    action = A2uiAction(
        name=user_action["name"],
        context=ctx if isinstance(ctx, dict) else {},
    )
    if isinstance(user_action.get("surfaceId"), str):
        action.surface_id = user_action["surfaceId"]
    if isinstance(user_action.get("sourceComponentId"), str):
        action.source_component_id = user_action["sourceComponentId"]
    return action


def _read_ai37_metadata(md: dict[str, Any], data: dict[str, Any]) -> Ai37Metadata:
    """metadata.ai37 может прийти в message.metadata, data.ai37 или data.metadata.ai37."""
    from_msg = md.get("ai37") if isinstance(md.get("ai37"), dict) else None
    nested = None
    if isinstance(data.get("metadata"), dict):
        nested = data["metadata"].get("ai37")
    from_data = data.get("ai37") if isinstance(data.get("ai37"), dict) else nested
    # message.metadata.ai37 перекрывает data.ai37 (как в TS).
    merged: dict[str, Any] = {**(from_data or {}), **(from_msg or {})}
    return _to_ai37_metadata(merged)


def _to_ai37_metadata(raw: dict[str, Any]) -> Ai37Metadata:
    context_files = raw.get("context_files")
    intent = raw.get("intent")
    return Ai37Metadata(
        tenant=raw.get("tenant"),
        app_id=raw.get("app_id"),
        channel=raw.get("channel"),
        thread_id=raw.get("thread_id"),
        session_id=raw.get("session_id"),
        context_refs=raw.get("context_refs"),
        context_files=(
            [_to_context_file(c) for c in context_files if isinstance(c, dict)]
            if isinstance(context_files, list)
            else None
        ),
        intent=(
            IntentEnvelope(skill=intent["skill"], params=intent.get("params"))
            if isinstance(intent, dict) and isinstance(intent.get("skill"), str)
            else None
        ),
        trace_id=raw.get("trace_id"),
        accepted_output_modes=raw.get("acceptedOutputModes"),
    )


def _to_context_file(raw: dict[str, Any]) -> ContextFile:
    return ContextFile(
        ref=raw.get("ref", ""),
        name=raw.get("name", ""),
        scope=raw.get("scope", "chat"),
        summary=raw.get("summary"),
        is_large=raw.get("isLarge"),
    )
